arangod/Replication2/ReplicatedLog/LogCore.cpp

```diff
@@ -0,0 +1,85 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2021-2021 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Tobias Gödderz
+////////////////////////////////////////////////////////////////////////////////
+
+#include "LogCore.h"
+
+#include "Replication2/ReplicatedLog/PersistedLog.h"
+
+#include <Basics/Exceptions.h>
+#include <Basics/debugging.h>
+#include <Basics/system-compiler.h>
+#include <Basics/voc-errors.h>
+
+#include <utility>
+
+using namespace arangodb;
+using namespace arangodb::replication2;
+using namespace arangodb::replication2::replicated_log;
+
+replicated_log::LogCore::LogCore(std::shared_ptr<PersistedLog> persistedLog)
+    : _persistedLog(std::move(persistedLog)) {
+  if (ADB_UNLIKELY(_persistedLog == nullptr)) {
+    TRI_ASSERT(false);
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
+                                   "When instantiating ReplicatedLog: "
+                                   "persistedLog must not be a nullptr");
+  }
+}
+
+auto replicated_log::LogCore::removeBack(LogIndex first) -> Result {
+  std::unique_lock guard(_operationMutex);
+  return _persistedLog->removeBack(first);
+}
+
+auto replicated_log::LogCore::insert(PersistedLogIterator& iter, bool waitForSync) -> Result {
+  std::unique_lock guard(_operationMutex);
+  PersistedLog::WriteOptions opts;
+  opts.waitForSync = waitForSync;
+  return _persistedLog->insert(iter, opts);
+}
+
+auto replicated_log::LogCore::read(LogIndex first) const -> std::unique_ptr<PersistedLogIterator> {
+  std::unique_lock guard(_operationMutex);
+  return _persistedLog->read(first);
+}
+
+auto replicated_log::LogCore::insertAsync(std::unique_ptr<PersistedLogIterator> iter, bool waitForSync)
+    -> futures::Future<Result> {
+  std::unique_lock guard(_operationMutex);
```

Comment:
As far as I understand this will aqcuire the mutex synchronously and only release it once the operation is actually done, since the guard is moved into the closure which is then attached to the future. Therefore, when multiple `insertAsync` calls are issued, the operation will only actually become "asynchronous", when the contention and waiting for other operations to finish is completed. Is this what we want here?

Reply:
Damn, good catch. 🤔  The `LogCore` is held as `unique_ptr` and there should only ever be a single user. The current code design (outside the core) ensures that there will be no other call to `insertAsync` before the previous one returned. The mutex is there to prevent the following scenario: if a leader issues a async insert operation (A) and then resigns, the log core is passed on to the new follower instance. If the new follower issues an async insert operation (B) due to an appendEntries request, we have to make sure that the operation (A) is completed before (B) starts. For performance reasons we write concurrently to rocksdb without using conflict detection. If two inserts run concurrently, this could break the on disk state.

I wonder what the correct behavior should be, @goedderz. I think the current behavior is what we actually need but calling the function `insertAsync` could be confusing.